helix-view/src/view.rs

```diff
@@ -176,84 +193,132 @@ impl View {
-                self.offset.row + line - (last_line.saturating_sub(scrolloff))
-            } else if line < self.offset.row + scrolloff {
-                // scroll up
-                line.saturating_sub(scrolloff)
+    ) -> Option<ViewPosition> {
+        let doc_text = doc.text().slice(..);
+        let viewport = self.inner_area(doc);
+        let vertical_viewport_end = self.offset.vertical_offset + viewport.height as usize;
+        let text_fmt = doc.text_format(viewport.width, None);
+        let annotations = self.text_annotations(doc, None);
+
+        // - 1 so we have at least one gap in the middle.
+        // a height of 6 with padding of 3 on each side will keep shifting the view back and forth
+        // as we type
+        let scrolloff = scrolloff.min(viewport.height.saturating_sub(1) as usize / 2);
+
+        let cursor = doc.selection(self.id).primary().cursor(doc_text);
+        let mut offset = self.offset;
+
+        let (visual_off, mut at_top) = if cursor >= offset.anchor {
+            let off = visual_offset_from_anchor(
+                doc_text,
+                offset.anchor,
+                cursor,
+                &text_fmt,
+                &annotations,
+                vertical_viewport_end,
+            );
+            (off, false)
+        } else if CENTERING {
+            // cursor out of view
+            return None;
+        } else {
+            (None, true)
+        };
+
+        let new_anchor = match visual_off {
+            Some((visual_pos, _)) if visual_pos.row < scrolloff + offset.vertical_offset => {
+                if CENTERING && visual_pos.row < offset.vertical_offset {
+                    // cursor out of view
+                    return None;
+                }
+                at_top = true;
+                true
+            }
+            Some((visual_pos, _)) if visual_pos.row >= vertical_viewport_end - scrolloff => {
+                if CENTERING && visual_pos.row >= vertical_viewport_end as usize {
+                    // cursor out of view
+                    return None;
+                }
+                true
+            }
+            Some(_) => false,
+            None => true,
+        };
+
+        if new_anchor {
+            if at_top {
+                (offset.anchor, offset.vertical_offset) = char_idx_at_visual_offset(
+                    doc_text,
+                    cursor,
+                    -(scrolloff as isize),
+                    0,
+                    &text_fmt,
+                    &annotations,
+                );
             } else {
-                self.offset.row
-            };
+                (offset.anchor, offset.vertical_offset) = char_idx_at_visual_offset(
+                    doc_text,
+                    cursor,
+                    -(viewport.height as isize - scrolloff as isize),
+                    0,
+                    &text_fmt,
+                    &annotations,
+                );
+            }
```

Comment:
seems like the third parameter is the only difference here, could we move the rest of the call after the if?

Reply:
Great point, that makes that code much more readable :) Fixed in [1ec785b](https://github.com/helix-editor/helix/pull/5420/commits/1ec785b0c6ae30dbee4251575dd33b286bb86dd8)